### DISCUSSION TITLE
What's the rationale for "use client"?

### DISCUSSION THREAD

**Discussion by shahyar** [irsyadadl/intentui#240] [General]
It seems to me that the majority of the components could in fact be rendered in SSR. What's the reason for preventing this with `"use client"`? Even stuff that generally depends on `window` or `document` could likely be rendered in SSR and then bound on client render.

**Comment by irsyadadl** [comment]
What component for example?

**Reply by shahyar** [reply]
They are _all_ using "use client", which prevents them from being rendered on the server to hydrate the DOM. For example, `loader` doesn't use framer-motion, has no outside dependencies, and yet is still prevented from rendering in RSC.

**Reply by irsyadadl** [reply]
Did you know the ProgressBar can't be place inside the server component? Loader need that, so we have to make it as client.

**Reply by shahyar** [reply]
Why not? I just tested `ProgressBar` from `react-aria-components` and it worked fine rendering to HTML.

**Reply by shahyar** [reply]
In fact, I render most of `react-aria-components` on the server without issue.

**Reply by irsyadadl** [reply]
Tell me what components then, so I can tweak in the major release of this #243 . List the RAC Component can run on server without flag `use client`.

**Reply by irsyadadl** [reply]
```tsx
import { ProgressBar } from "react-aria-components"

export default function Page() {
  return <ProgressBar />
}
```

Then you got 

```bash
TypeError: createContext only works in Client Components. Add the "use client" directive at the top of the file to use it. Read more: https://nextjs.org/docs/messages/context-in-server-component

This error happened while generating the page. Any console logs will be displayed in the terminal window.
Source
app/sink/page.tsx (1:1) @ [project]/app/sink/page.tsx [app-rsc] (ecmascript)

> 1 | import { ProgressBar } from "react-aria-components"
    | ^
  2 |
  3 | export default function Page() {
  4 |   return <ProgressBar />
```

**Reply by shahyar** [reply]
That's specifically an issue with Next. Vite/Vike do not have this restriction, and correctly let React render a context on the server.

**Reply by irsyadadl** [reply]
I’ve known about that for a long time. even on irsyad.co, i don’t use `use client` at all, and everything works perfectly fine with server components. the issue, though, is that most people today heavily rely on next.js as their front-end. we (or i) can’t be that selfish. honestly, i’m a laravel user, not next.js, but i can’t be selfish about that. we’ll let them choose whatever they want. and we both know that every editor has a `find & replace all` feature, which can be useful, right? 
 
in the next version, i'll remove `use client` when you add it through the cli if vite is detected. also, if you’re using laravel, when you add a single component or all components, it’s guaranteed that they no longer use the `use client` flag. just give it a try. i haven’t used anything more than laravel. so for now, the exception is only made for laravel. in the next version, the default will be to exclude that flag if vite is detected.

**Reply by irsyadadl** [reply]
@shahyar fixed on [v1.2.47]( https://github.com/justdlabs/cli/releases/tag/v1.2.47)
